src/index.ts

```diff
@@ -47,89 +136,53 @@ export class I18N {
+        let text: string | undefined;
+        let fallbackText: string | undefined;
+        let details: TranslationData['details'];
 
-        const keyset = languageData[keysetName];
-
-        if (!keyset) {
-            this.warn(
-                'Keyset not found.',
-                keysetName,
-            );
+        ({text, fallbackText, details} = this.getTranslationData({keysetName, key, params, lang: this.lang}));
 
-            return key;
+        if (details && details.code !== ErrorCode.NoLanguageData) {
+            const message = mapErrorCodeToMessage({
+                code: details.code,
+                lang: this.lang,
+                fallbackLang: this.fallbackLang,
+            });
+            this.warn(message, details.keysetName, details.key);
         }
 
-        if (Object.keys(keyset).length === 0) {
-            this.warn(
-                'Keyset is empty.',
-                keysetName,
-            );
-
-            return key;
-        }
-
-        const keyValue = keyset && keyset[key];
-        let result: string;
-
-        if (typeof keyValue === 'undefined') {
-            this.warn(
-                'Missing key.',
+        if (this.fallbackLang) {
+            ({text, fallbackText, details} = this.getTranslationData({
                 keysetName,
                 key,
-            );
-
-            return key;
-        }
-
-        if (Array.isArray(keyValue)) {
-            if (keyValue.length < 3) {
-                this.warn(
-                    'Missing required plurals',
-                    keysetName,
-                    key,
-                );
-
-                return key;
-            }
-
-            const count = Number(params?.count);
-
-            if (Number.isNaN(count)) {
-                this.warn(
-                    'Missing params.count for key.',
-                    keysetName,
-                    key,
-                );
-
-                return key;
-            }
-
-            const pluralizer = this.getLanguagePluralizer(this.lang);
-            result = keyValue[pluralizer(count, PluralForm)] || keyValue[PluralForm.Many];
-
-            if (keyValue[PluralForm.None] === undefined) {
-                this.warn('Missing key for 0', keysetName, key);
+                params,
+                lang: this.fallbackLang,
+            }));
+
+            if (details && details.code !== ErrorCode.NoLanguageData) {
+                const message = mapErrorCodeToMessage({
+                    code: details.code,
+                    lang: this.lang,
+                    fallbackLang: this.fallbackLang,
+                });
+                this.warn(message, details.keysetName, details.key);
             }
-        } else {
-            result = keyValue;
         }
 
-        if (params) {
-            result = replaceParams(result, params);
+        if (!text && !fallbackText) {
+            const message = mapErrorCodeToMessage({
+                code: details?.code,
```

Comment:
Is only `NoLanguageData` code possible here?

Reply:
Yep